audit/audit.go

```diff
@@ -0,0 +1,164 @@
+type keyValuePair struct {
+	Key   string `avro:"key" json:"key,omitempty"`
+	Value string `avro:"value" json:"value,omitempty"`
+}
+
+// OutboundProducer defines a producer for sending outbound audit events to a kafka topic
+type OutboundProducer interface {
+	Output() chan []byte
+}
+
+//AuditorService defines the behaviour of an audior
+type AuditorService interface {
+	GetEvent(input context.Context) (*Event, error)
+	Record(ctx context.Context, action string, result string, params Params) error
+}
+
+//Auditor provides functions for interception HTTP requests and populating the context with a base audit event and
+// recording audit events
+type Auditor struct {
+	namespace     string
+	tokenName     string
+	marshalToAvro avroMarshaller
+	producer      OutboundProducer
+}
+
+//New creates a new Auditor with the namespace, producer and token provided.
+func New(producer OutboundProducer, namespace string, token string) *Auditor {
+	return &Auditor{
+		producer:      producer,
+		namespace:     namespace,
+		tokenName:     token,
+		marshalToAvro: EventSchema.Marshal,
+	}
+}
+
+//Interceptor returns an http.Handler which populates a http.Request.Context with a base audit event - fields common
+// to all events regardless of context before passing it on to the handler in the chain.
+func (a *Auditor) Interceptor() func(handler http.Handler) http.Handler {
+	return func(handler http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx := r.Context()
+			event := Event{
+				RequestID: r.Header.Get(requestIDHeader),
+				Namespace: a.namespace,
+				Params:    make([]keyValuePair, 0),
+			}
+
+			if serviceToken := r.Header.Get(a.tokenName); len(serviceToken) > 0 {
+				ctx = context.WithValue(ctx, a.tokenName, serviceToken)
+				event.Service = serviceToken
+			}
+
+			if from := r.Header.Get(fromHeader); len(from) > 0 {
+				ctx = context.WithValue(ctx, fromContextKey, from)
+				event.User = from
+			}
+
+			ctx = context.WithValue(ctx, eventContextKey, event)
+			handler.ServeHTTP(w, r.WithContext(ctx))
+		})
+	}
+}
+
+//Record captures the provided action, result and parameters and an audit event. Common fields - time, user, service
+// are added automatically. An error is returned if there is a problem recording the event it is up to the caller to
+// decide what do with the error in these cases.
+func (a *Auditor) Record(ctx context.Context, action string, result string, params Params) error {
+	if action == "" {
+		return auditError("attempted action is required field", "nil", params)
+	}
+	if result == "" {
+		return auditError("result is required field", action, params)
+	}
+
+	e, err := a.GetEvent(ctx)
+	if err != nil {
+		return auditError(err.Error(), action, params)
+	}
+
+	e.AttemptedAction = action
+	e.Result = result
+
+	if params != nil {
+		for k, v := range params {
+			e.Params = append(e.Params, keyValuePair{Key: k, Value: v})
+		}
+	}
+
+	e.Created = time.Now().String()
```

Comment:
This will create a string which includes any current TZ offset (e.g. BST on dev machines, but likely UTC on the servers) - do we want to enforce UTC for all logging for consistency?

I think UTC times are more commonly used in other logging.

Perhaps:  `e.Created = time.Now().UTC().Format(time.RFC3339Nano)`